Fully supports progressive JPGS and Interlaced images for 3.x

konnng-dev opened the pull request:
The intent of this PR is to introduce support to generate progressive JPEGs or interlaced images.

Couple days ago, the author of Intervention provided support to accomplish that https://github.com/Intervention/image/releases/tag/3.6.0

With that, I did some adjustments on Glide API to get the desired support.

**Highlights:**

* `Encode` manipulator moved from the `$manipulators` list, because it needs to be the last manipulator to run when generating the images.
* `Encode` manipulator now returns a `EncodedImage`. This change avoids to double encode image when generating it.
* `Encode` also makes more use of Intervention `enums` to encode the image properly.
* interlace parameter also was added to create progressive JPEGs or interlaced PNGs or GIFs.

konnng-dev commented:
TODO: work on tests to check for progressive / interlaced images.

konnng-dev commented:
@ADmad would you mind to take a look and tell me what you think? 

I'm not sure if introducing this new encoding strategy would be the best for this package.

feedbacks are welcome :)

ADmad commented:
Sorry I haven't had the time to look into this but could you please address the failing tests in the meantime.

konnng-dev commented:
@ADmad I addressed all the issues that Github Actions was issuing.

For testing, could you point me the right place where I could do the test for the  progressive/interlaced images I am introducing?

konnng-dev commented on src/Manipulators/Encode.php:
```diff
@@ -15,30 +12,39 @@ class Encode extends BaseManipulator
         }
 
-        return $image;
+        return $image->encodeByExtension(...$encoderOptions);
```
Due the latest changes on Intervention package. It was necessary to change the return type of this class, to return the encoded image instance instead just the image instance.

konnng-dev commented on src/Api/Api.php:
```diff
@@ -91,10 +92,15 @@ public function run(string $source, array $params): string
         }
 
-        return $image->encodeByMediaType()->toString();
+        $encoder = new Encode();
+        $encoder->setParams($params);
+        /** @var \Intervention\Image\Interfaces\EncodedImageInterface */
+        $encoded = $encoder->run($image);
+
+        return $encoded->toString();
```
Since there was some important changes on how Intervention process images, it was necessary to move the Encode manipulator to run here instead. 

This will make sure that it will run at the end of all manipulators, sine it returns a `EncodedImage` instance. Also it will avoid the strategy used before, causing double encoding of the given image.

konnng-dev replied:
Ok, makes sense.

I moved the Encoder file to the Api folder, just to keep SoC and not pollute Api.php file.

konnng-dev commented:
@ADmad I added some comments to help the review process. 

Ideally, it would be good to add tests for the new option added to the image API. But I don't have idea what would be the best place to put it.

konnng-dev commented on src/Manipulators/Encode.php:
```diff
@@ -15,30 +12,39 @@ class Encode extends BaseManipulator
+                break;
+            case 'gif':
+            case 'png':
+                $encoderOptions['interlaced'] = $shouldInterlace;
```
Unfortunately, there is no such option anymore.

* https://github.com/Intervention/image/blob/develop/src/Encoders/GifEncoder.php#L11
* https://github.com/Intervention/image/blob/develop/src/Encoders/PngEncoder.php#L11

Options are pretty limited in this new version.

konnng-dev commented:
@ADmad I applied the suggested changes and also add a new doc entry for the new option. I'm just not sure how to preview documentation locally, but I guess it is following the same convention used before.

Also, do you have idea on how we could test this new feature? I just updated `EncodeTest` to comply with the recent changes, but I didn't have idea on how test that.

konnng-dev commented:
@ADmad friendly ping to check if you had time to review changes.

ADmad commented on tests/Api/EncodeTest.php:
```diff
@@ -0,0 +1,265 @@
+     *
+     * @psalm-suppress PropertyNotSetInConstructor
+     */
+    private $encoder;
```
You specify the type for all these properties and avoid the `@var` tags.

ADmad commented on tests/Api/EncodeTest.php:
```diff
@@ -0,0 +1,265 @@
+            $manager->create(100, 100)->encode(new MediaTypeEncoder('image/gif'))->toFilePointer()
+        );
+
+        if (function_exists('imagecreatefromwebp')) {
```
Shouldn't this function be always available for PHP 8.1+?

ADmad commented on tests/Api/EncodeTest.php:
```diff
@@ -0,0 +1,265 @@
+            );
+        }
+
+        if (function_exists('imagecreatefromavif')) {
```
Shouldn't this function be always available for PHP 8.1+?

ADmad commented on tests/Api/EncodeTest.php:
```diff
@@ -0,0 +1,265 @@
+        /**
+         * @psalm-suppress ArgumentTypeCoercion
+         */
+        $this->assertInstanceOf('League\Glide\Api\Encode', $this->encoder);
```
```suggestion
        $this->assertInstanceOf(Encode:class, $this->encoder);
```
with `use` statement for it.

ADmad commented on tests/Api/EncodeTest.php:
```diff
@@ -0,0 +1,265 @@
+        $this->assertSame('image/jpeg', $this->getMime($this->encoder->setParams(['fm' => 'jpg'])->run($this->jpg)));
+        $this->assertSame('image/jpeg', $this->getMime($this->encoder->setParams(['fm' => 'jpg'])->run($this->png)));
+        $this->assertSame('image/jpeg', $this->getMime($this->encoder->setParams(['fm' => 'jpg'])->run($this->gif)));
+        $this->assertSame('image/jpeg', $this->getMime($this->encoder->setParams(['fm' => 'pjpg'])->run($this->jpg)));
```
Do we have a way to check if the image is actually interlaced?

ADmad commented on docs/3.0/api/progressive-interlaced.md:
```diff
@@ -0,0 +1,27 @@
+
+The `interlaced` parameter controls whether an image is rendered in a progressive or interlaced format. This feature enhances the loading experience of images, making them appear gradually as they are downloaded, which can improve the user experience on slower connections.
+
+> Caution: for GIF/PNG, it can generate a sligher larger file size.
```
```suggestion
> Caution: for GIF/PNG, it can generate a slightly larger file size.
```

ADmad commented on docs/3.0/api/progressive-interlaced.md:
```diff
@@ -0,0 +1,27 @@
+- **JPG**: The `Interlaced` parameter applies a progressive scan to JPG images.
+- **PNG** and **GIF**: The `Interlaced` parameter enables interlacing for GIF/PNG images.
+
+> Note: Case `ext` is set to `.pjpg`, it will automatically generate a progressive JPG image, regardless of the `interlaced` parameter.
```
```suggestion
> Note: In case the `fm` parameter is set to `.pjpg`, it will automatically generate a progressive JPG image, regardless of the `interlaced` parameter.
```

ADmad reviewed (commented):
A issue with having the encoder as a separate step instead of a manipulation is that one cannot add manipulators to be run after the image is encoded. For e.g. there's this [optimizer](https://github.com/YannikFirre/glide-optimizer) manipulator which needs to run after encoding.

C2H6-383 commented:
Just wanted to check up on the progress of this pr: Are there any updates about the progression to merge this into 3.x?

I would really like to see version 3 to be released but afaik the release depends on the merge of this pr...

konnng-dev commented:
@C2H6-383 I ended up pushing the asked changes to this PR. I appreciate the reminder. 

I will try to address it this week so @ADmad can reviiew it.

C2H6-383 commented:
Hey @konnng-dev, just wanted to check up on this issue again. Are there more updates on this PR to allow it to be merged? 😅

ADmad commented:
Since @konnng-dev seems to be unavailable I'll try and finish this up myself.

juliovedovatto commented:
Thanks for giving proper attention @ADmad  and my apologies for overpromising. I ended up delivering the project using this package for the client and moved on to a different projects. Which caused me to sidetrack and lost focus on delivering this one.

I hope my updates helps this library to be better compatible with Intervention v3.

konnng-dev commented:
Thanks for giving proper attention @ADmad and my apologies for overpromising. I ended up delivering the project using this package for the client and moved on to a different projects. Which caused me to sidetrack and lost focus on delivering this one.

I hope my updates helps this library to be better compatible with Intervention v3.

ADmad commented:
@konnng-dev No apology needed, thank you for your contributions.
